src/cheri_insts.sail

```diff
@@ -666,19 +672,50 @@ function clause execute (CSeal(cd, cs, ct)) =
 union clause ast = CCSeal : (regidx, regidx, regidx)
 function clause execute (CCSeal(cd, cs, ct)) =
 {
-  let cs_val = readCapReg(cs);
-  let ct_val = readCapReg(ct);
+  let cs_val = readCapReg(cs); /* Tagged but unsealed cap */
+  let ct_val = readCapReg(ct); /* Authority */
   let ct_cursor = getCapCursor(ct_val);
   let (ct_base, ct_top) = getCapBounds(ct_val);
+
   if not (cs_val.tag) then {
     handle_cheri_reg_exception(CapEx_TagViolation, cs);
     RETIRE_FAIL
-  } else if not (ct_val.tag) | ct_val.address == to_bits(sizeof(xlen), otype_unsealed) then {
-    writeCapReg(cd, cs_val);
-    RETIRE_SUCCESS
   } else if cs_val.sealed then {
+    /*
+     * If the input to be conditionally sealed is already sealed, just pass it
+     * through before any other checks are made.
+     */
     handle_cheri_reg_exception(CapEx_SealViolation, cs);
     RETIRE_FAIL
+  } else if ct_val.address == to_bits(sizeof(xlen), otype_unsealed) then {
+    /*
+     * If the request is to seal to the unsealed type, just pass the result
+     * through.  The authority is probably the result of CCopyType from an
+     * unsealed capability and so cs_val is probably unsealed.
+     */
+    writeCapReg(cd, cs_val);
+    RETIRE_SUCCESS
+  } else if not (ct_val.tag) then {
+    /*
+     * An untagged authority can construct sealed capabilities at reserved
+     * otypes if those actions are ambiently available.  Otherwise, this likely
+     * represents an authority taken so far out of bounds by CCopyType as to be
+     * unrepresentable.  In that case, the caller may hold another authority
+     * they wish to try, so just pass the result through.
+     */
+    if ct_cursor == otype_sentry then {
+      let (success, newCap) = sealCap(cs_val, to_bits(otype_width, ct_cursor));
+      if not (success) then {
+        handle_cheri_reg_exception(CapEx_InexactBounds, cs);
+        RETIRE_FAIL
+      } else {
+        writeCapReg(cd, cs_val);
+        RETIRE_SUCCESS
+      }
+    } else {
+      handle_cheri_reg_exception(CapEx_TagViolation, cs); // XXX?
```

Comment:
Shouldn't this fall-through case be writeCapReg(cd, cs_val); RETIRE_SUCCESS as before? Presumably this is just copied from CSeal, or is this an intended behaviour change?

Reply:
Ah, yes, right, this probably should just fall through as before.  Don't know how I got my wires crossed; maybe it was, indeed, copy-paste.  Will fix.